From 6bc7acbedee1e23d8cdbe110ad31c514e23e7131 Mon Sep 17 00:00:00 2001
From: Dionysios Kalofonos <peitemou@gmail.com>
Date: Mon, 17 Mar 2014 13:33:59 +0200
Subject: [PATCH] Fixes some memory leaks.

---
 Source/CTCoreMessage.m | 29 ++++++++++++++++++++++++-----
 1 file changed, 24 insertions(+), 5 deletions(-)

diff --git a/Source/CTCoreMessage.m b/Source/CTCoreMessage.m
index bdd2cc8..8d25c75 100644
--- a/Source/CTCoreMessage.m
+++ b/Source/CTCoreMessage.m
@@ -89,10 +89,11 @@
     if (myMessage != NULL) {
         mailmessage_flush(myMessage);
         mailmessage_free(myMessage);
+        if (myFields)
+            mailimf_single_fields_free(myFields);
     }
-    if (myFields != NULL) {
-        mailimf_single_fields_free(myFields);
-    }
+    else
+        [self _releaseMailimfSingleFields: myFields];
     self.lastError = nil;
     self.parentFolder = nil;
     [myParsedMIME release];
@@ -141,8 +142,7 @@
 }
 
 - (void)setFields:(struct mailimf_fields *)fields {
-    if (myFields != NULL)
-        mailimf_single_fields_free(myFields);
+    [self _releaseMailimfSingleFields: myFields];
     myFields = mailimf_single_fields_new(fields);
 }
 
@@ -867,4 +867,23 @@
 }
 
 
+- (void)_releaseMailimfSingleFields:(struct mailimf_single_fields *)fields {
+    if (fields) {
+        if (fields->fld_bcc) mailimf_bcc_free(fields->fld_bcc);
+        if (fields->fld_cc) mailimf_cc_free(fields->fld_cc);
+        if (fields->fld_comments) mailimf_comments_free(fields->fld_comments);
+        if (fields->fld_from) mailimf_from_free(fields->fld_from);
+        if (fields->fld_in_reply_to) mailimf_in_reply_to_free(fields->fld_in_reply_to);
+        if (fields->fld_keywords) mailimf_keywords_free(fields->fld_keywords);
+        if (fields->fld_message_id) mailimf_message_id_free(fields->fld_message_id);
+        if (fields->fld_orig_date) mailimf_orig_date_free(fields->fld_orig_date);
+        if (fields->fld_references) mailimf_references_free(fields->fld_references);
+        if (fields->fld_reply_to) mailimf_reply_to_free(fields->fld_reply_to);
+        if (fields->fld_sender) mailimf_sender_free(fields->fld_sender);
+        if (fields->fld_subject) mailimf_subject_free(fields->fld_subject);
+        if (fields->fld_to) mailimf_to_free(fields->fld_to);
+        mailimf_single_fields_free(fields);
+    }
+}
+
 @end
-- 
1.8.5.2 (Apple Git-48)

